spec/client_spec.rb

```diff
@@ -76,6 +76,55 @@ def redis.with
     context 'when lock is available' do
       after(:each) { lock_manager.unlock(@lock_info) if @lock_info }
 
+      context 'when redis connection error occurs' do
+        let(:servers_with_quorum) {
+          [
+            "redis://#{redis1_host}:#{redis1_port}",
+            "redis://#{redis2_host}:#{redis2_port}",
+            unreachable_redis
+          ]
+        }
+
+        let(:servers_without_quorum) {
+          [
+            "redis://#{redis1_host}:#{redis1_port}",
+            "redis://127.0.0.1:3124",
+            unreachable_redis
+          ]
+        }
+
+        it 'locks if majority of redis instances are available' do
+          redlock = Redlock::Client.new(servers_with_quorum)
+
+          expect {
+            expect(redlock.lock(resource_key, ttl)).to be_truthy
+          }.to_not raise_error
+
+        end
+
+        it 'fails to acquire a lock if majority of Redis instances are not available' do
+          redlock = Redlock::Client.new(servers_without_quorum)
+
+          expect {
+            redlock.lock(resource_key, ttl)
+          }.to raise_error Redlock::LockAcquisitionError
+        end
+
+        it 'recovers from connection errors when server comes back' do
```

Comment:
this test is flawed — it's hard to tell whether anything recovers when you're constructing Redlock client from scratch before checking whether it'll work with `servers_with_quorum`.. it has to be the same instance with underlying Redis instances behaving differently. this can be mocked like so:

```ruby
before do
  expect(some_redis_instance).to receive(:some_lock_call).and_raise(StandardError).ordered
  expect(some_redis_instance).to receive(:some_lock_call).and_call_original.ordered
```

maybe adding `.once` or `.times(1)` would also be benefitial here (composed the code without refering to the docs)

Reply:
alternatively you could leave old test as it was here before:

```ruby
    context 'when a server comes back' do
      it 'recovers from connection issues' do
        # Same as above.
        redis_instance = lock_manager.instance_variable_get(:@servers).first
        old_redis = redis_instance.instance_variable_get(:@redis)
        redis_instance.instance_variable_set(:@redis, unreachable_redis)
        expect {
          lock_manager.lock(resource_key, ttl)
        }.to raise_error(RedisClient::CannotConnectError)
        redis_instance.instance_variable_set(:@redis, old_redis)
        expect(lock_manager.lock(resource_key, ttl)).to be_truthy
      end
    end
```

Reply:
Makes sense, not sure what was my initial intention when I was removing these tests 😕 Re-added and slightly adjusted the tests that were removed